src/osdc/Journaler.h

```diff
@@ -529,76 +529,76 @@ class Journaler {
   // ===================
 
   Header get_last_committed() const {
-    ceph_assert(!ceph_mutex_is_locked_by_me(lock));
+    ceph_assert(!ceph::mutex_debugging || !ceph_mutex_is_locked_by_me(lock));
```

Comment:
Are these asserts actually useful/worth it?  In all cases they are followed by `lock_guard l(lock);` which if the mutex is already held by the same thread would throw `std::system_error` or just deadlock (as implementations are only encouraged to detect invalid usage).  Either way, it's not hard to diagnose.

There is only one existing instance of such an assert (in `src/common/config_proxy.h`), so it's almost unprecedented too.

Reply:
> Are these asserts actually useful/worth it? In all cases they are followed by `lock_guard l(lock);` which if the mutex is already held by the same thread would throw `std::system_error` or just deadlock (as implementations are only encouraged to detect invalid usage). Either way, it's not hard to diagnose.

I assume you're citing from: https://en.cppreference.com/w/cpp/thread/mutex/lock

That document says the behavior is undefined so I would say the assert is warranted.

> There is only one existing instance of such an assert (in `src/common/config_proxy.h`), so it's almost unprecedented too.

Yes, I added that too. It has been very helpful in my experience for debugging.

Reply:
> I assume you're citing from: https://en.cppreference.com/w/cpp/thread/mutex/lock

does [ceph::mutex_debug::lock()](https://github.com/ceph/ceph/blob/main/src/common/mutex_debug.h#L181) not do this checking when mutex debugging is enabled?

Reply:
Possibly, but let's not delay getting this merged if it fixes the problem.

Reply:
https://github.com/ceph/ceph/pull/60037